fs/hayleyfs/h_dir.rs

```diff
@@ -95,119 +95,85 @@ impl<'a, State, Op> DentryWrapper<'a, State, Op> {
 }
 
 impl<'a> DentryWrapper<'a, Clean, Recovering> {
-    pub(crate) fn get_recovering_dentry(info: DentryInfo) -> Result<Self> {
+    pub(crate) fn from_dinfo(info: DentryInfo) -> Result<Self> {
         // use the virtual address in the DentryInfo to look up the
         // persistent dentry
         let dentry: &mut HayleyFsDentry =
             unsafe { &mut *(info.get_virt_addr().unwrap() as *mut HayleyFsDentry) };
-
-        if dentry.ino == 0 {
-            pr_info!("ERROR: dentry is invalid\n");
-            return Err(EPERM);
-        };
         Ok(Self {
             state: PhantomData,
             op: PhantomData,
             dentry,
         })
     }
 
-    pub(crate) fn rename_ptr(&self, sbi: &SbInfo) -> Option<&'a mut HayleyFsDentry> {
+    pub(crate) fn rename_ptr(&self, sbi: &SbInfo) -> Result<Option<&'a mut HayleyFsDentry>> {
         if self.dentry.rename_ptr == 0 {
-            None
+            Ok(None)
         } else {
             let base = sbi.get_virt_addr() as * const u8;
             let off: isize = self.dentry.rename_ptr.try_into().unwrap();
-            let ptr = unsafe { base.offset(off) } as *mut HayleyFsDentry;
-            unsafe { ptr.as_mut() }
+
+            let pg_sz: isize = HAYLEYFS_PAGESIZE.try_into().unwrap();
+            let start = pg_sz * (sbi.get_data_pages_start_page() as isize);
+            let end = pg_sz * (sbi.get_size() as isize);
+
+            if off < start || off >= end {
+                Err(ESPIPE) // XXX: ?
+            } else {
+                let ptr = unsafe { base.offset(off) } as *mut HayleyFsDentry;
+                Ok(unsafe { ptr.as_mut() })
+            }
         }
     }
 
-    //
-    
-    pub(crate) fn get_dentry(&mut self) -> &'_ mut HayleyFsDentry {
-        self.dentry
-    }
 
-    fn clear_rename_pointer(self) -> Result<DentryWrapper<'a, Dirty, Recovering>> {
-        self.dentry.rename_ptr = 0;
-        
-        Ok(DentryWrapper {
-            state: PhantomData,
-            op: PhantomData,
-            dentry: self.dentry,
-        })
+    pub(crate) fn into_init_rename(self, sbi: &SbInfo) 
+        -> Result<DentryWrapper<'a, Clean, InitRenamePointer>> {
```

Comment:
I don't love that `InitRenamePointer` is used here as a catch-all for any directory entry with a rename pointer, since it has a more specific meaning in regular execution. I think this implementation is correct, though, so let's leave it as is and I'll make a note to investigate further later on.